src/docfx/publish/PublishModelBuilder.cs

```diff
@@ -1,28 +1,31 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System;
 using System.Collections.Concurrent;
 using System.Collections.Generic;
+using System.Diagnostics;
 using System.IO;
 using System.Linq;
 
 namespace Microsoft.Docs.Build
 {
     internal class PublishModelBuilder
     {
-        public const string NonVersion = "NONE_VERSION";
-
+        private readonly object _lock = new object();
+        private readonly Context _context;
         private readonly string _outputPath;
         private readonly Config _config;
         private readonly Output _output;
         private readonly ErrorLog _errorLog;
-        private readonly ConcurrentDictionary<string, ConcurrentBag<FilePath>> _outputPathConflicts = new ConcurrentDictionary<string, ConcurrentBag<FilePath>>(PathUtility.PathComparer);
-        private readonly ConcurrentDictionary<string, ConcurrentDictionary<FilePath, IReadOnlyList<string>>> _filesBySiteUrl = new ConcurrentDictionary<string, ConcurrentDictionary<FilePath, IReadOnlyList<string>>>(PathUtility.PathComparer);
-        private readonly ConcurrentDictionary<string, FilePath> _filesByOutputPath = new ConcurrentDictionary<string, FilePath>(PathUtility.PathComparer);
-        private readonly ConcurrentDictionary<FilePath, PublishItem> _publishItems = new ConcurrentDictionary<FilePath, PublishItem>();
+        private readonly ConcurrentDictionary<PublishItemKey, Lazy<(FilePath, PublishItem, ConflictingType, ConcurrentDictionary<FilePath, IReadOnlyList<string>>)>> _outputConflicts
```

Comment:
The key could still be `PublishItem`, as long as the class is `PublishItemDuplicateComparer : IEqualityComparer<PublishItem>`

```suggestion
        private readonly ConcurrentDictionary<PublishItemKey, Lazy<(FilePath, PublishItem, ConflictingType, ConcurrentDictionary<FilePath, IReadOnlyList<string>>)>> _outputConflicts
```